bzerolib/connection/messenger/signalr/signalr.go

```
@@ -0,0 +1,328 @@
+	}
+
+	// Build our Url
+	u, err := buildUrl(targetUrl, params)
+	if err != nil {
+		return err
+	}
+
+	// Connect to our endpoint
+	if err := s.client.Dial(u, ctx); err != nil {
+		return fmt.Errorf("failed to connect to endpoint %s: %w", u.String(), err)
+	}
+
+	// Negotiate our SignalR version
+	// Ref: https://stackoverflow.com/questions/65214787/signalr-websockets-and-go
+	versionMessageBytes := append([]byte(`{"protocol": "json","version": 1}`), signalRMessageTerminatorByte)
+	if err := s.client.Send(versionMessageBytes); err != nil {
+		s.client.Close()
+		return fmt.Errorf("failed to negotiate SignalR version: %w", err)
+	}
+
+	// If the handshake was successful, then we've made our connection and we can
+	// start listening and sending on it
+	s.tmb.Go(func() error {
+		defer s.logger.Info("SignalR processing done")
+		defer close(s.doneChan)
+
+		// Unwrap and forward inbound messages
+		for {
+			select {
+			case <-s.tmb.Dying(): // death from Close() call
+				err := s.avengersEndgame()
+				if err != nil {
+					s.logger.Error(err)
+				}
+
+				s.client.Close()
+				return err
+			case <-s.client.Done():
+				return fmt.Errorf("closed websocket")
+			case messageBytes := <-s.client.Inbound():
+				if err := s.unwrap(*messageBytes); err != nil {
+					s.logger.Errorf("error unwrapping SignalR message: %s", err)
+				}
+			}
+		}
+	})
+	return nil
+}
+
+func (s *SignalR) avengersEndgame() error {
+	finalCountdown := time.NewTicker(time.Minute)
+	checkDoneInterval := time.Second
+
+	for {
+		select {
+		// The listener is probably not listening but we need to match completion messages
+		case messageBytes := <-s.client.Inbound():
+			if err := s.unwrap(*messageBytes); err != nil {
+				s.logger.Errorf("error unwrapping SignalR message: %s", err)
+			}
+		case <-time.After(checkDoneInterval):
+			if s.invocator.IsEmpty() {
+				return nil
+			}
+		case <-finalCountdown.C:
+			return fmt.Errorf("connection failed to close, forcing shutdown")
+		}
+	}
```

If the only purpose of this is to keep unwrapping signalR message so we remove them from invocator's tracked messages but then drop the messages anyway- why not just call a method on the invocator to clear all tracked messages instead?

---

This comment is inaccurate but regardless we're not dropping these messages

---

Ah right of course `unwrap` is still pushing the messages onto the inbound queue. Ok makes sense so this is to allow graceful shutdown where we still allow reading from all the messages in the `s.client.Inbound()` queue. Ok so that brings up another question should we still allow sending more messages while we are in this state? Maybe the `send` method should return an error if `! s.tmb.Alive()`

---

I'm going to think about this more but I don't know if we want that. My first instinct is that I'd rather be more cautious here and send more messages.

---

Alright thats fair. We have the `finalCountdown` anyway that is going to stop receiving the messages.